P4 seqflags

anwestm opened the pull request:
+ saves sequence flags every 16th messages

(asic launch config on startup is in p1 branch)

anwestm committed:
v1

anwestm committed:
Merge branch 'master' of https://github.com/tstana/cubes-fw

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

anwestm committed:
c2

anwestm committed:
- saves sequence flags every 16 messages
- moved code for setting temporary voltage from main

anwestm committed:
- removed coded for asic configurations which is already implemeted in subproject 1's branch

tstana commented on mem_mgmt/mem_mgmt.c:
```diff
@@ -158,7 +158,7 @@ nvm_status_t nvm_save_msp_seqflags(void)
 	if((NVM_SUCCESS == status)||(NVM_WRITE_THRESHOLD_WARNING == status)){
 		status = NVM_write(NVM_SEQFLAG_ADDR,
-		                   (uint8_t*)&seqflags,
+		                   (uint32_t*)&seqflags,
```
I get a warning from the compiler that this is an incompatible pointer type. The previous cast should have been OK, but I forget now why I said this should be changed to a `uint32_t` pointer...?

anwestm replied:
We talked about changing the cast in the function nvm_restore_msp_seqflags since the msp_seqflags union is 32 bits. I have  changed wrong part of the code here. Now changed back to uint8_t

tstana commented on mem_mgmt/mem_mgmt.c:
```diff
@@ -168,7 +168,7 @@ nvm_status_t nvm_save_msp_seqflags(void)
 {
 	msp_seqflags_t seqflags;
-	NVM_write(0xF000, 0, sizeof(msp_seqflags_t), NVM_DO_NOT_LOCK_PAGE);
+	//NVM_write(0xF000, 0, sizeof(msp_seqflags_t), NVM_DO_NOT_LOCK_PAGE);
```
This line can be removed. No idea why it was added, I see no point in clearing the flags in the NVM before reading them.

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -664,6 +665,10 @@ void msp_exprecv_complete(unsigned char opcode)
+	if (++msg_acc >= SEQ_FLAG_SAVE_INTERVAL) {
+		msg_acc = 0;
+		nvm_save_msp_seqflags();
+	}
```
This should be done on all type of transactions, not only MSP receive transactions. Suggest you move this code into a function, which you call from the `_complete` function of all type of transactions.

tstana commented on mem_mgmt/mem_mgmt.h:
```diff
@@ -44,10 +44,12 @@
+#define RAM_ADDR        	(0x20000000u)
+#define NVM_ADDR        	(0x60000000u)
+#define CITIROC_OFFSET  	(0xF010u)
+#define SEQFLAG_OFFSET  	(0xF100)
```
Please revert these four lines to their previous state, one TAB to the left, to avoid merge conflicts when we merge with Patrik's and Joar's code.

tstana reviewed (changes requested):
Thanks for these changes @anwestm. I've added some comments for you to address in the code.

anwestm committed:
- code cleanup
- changed 32 bit cast to 8 bit cast
- sequency flag accumulator now moved to function

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -738,6 +734,22 @@ void msp_exprecv_syscommand(unsigned char opcode)
+ * Accumulates a counter and saves sequence flags to NVM at an interval defined
+ * by SEQ_FLAG_SAVE_INTERVAL
+ */
+void accumulate_msp_seq_flag_save(void)
```
This function should be made `static` (so it's file-local) and moved to the top of the file, for example right below `prep_payload_data()`. Without the move to the top of the file, there are warnings for undefined reference to this function -- and the possibility for crashes when the code is run.

It also needs to be called as part of `msp_exprecv_syscommand()`.

anwestm committed:
+moved function and made static
+added seq flag save to sys commands

tstana reviewed (approved):
Thanks for your changes @anwestm. It looks good now.
